activerecord/lib/active_record/future_result.rb

```diff
@@ -0,0 +1,96 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  class FutureResult
+    Canceled = Class.new(ActiveRecordError)
+
+    delegate :empty?, :to_a, to: :result!
+
+    def initialize(pool, *args, **kwargs)
+      @mutex = Mutex.new
+
+      @pool = pool
+      @args = args
+      @kwargs = kwargs
+
+      @executed = false
+      @error = nil
+      @result = nil
+    end
+
+    def schedule!
+      ActiveRecord::Base.asynchronous_queries_tracker.register(self)
+      @pool.schedule_query(self)
+    end
+
+    def execute!(connection)
+      execute_query(connection)
+    end
+
+    def cancel!
+      @executed = true
+      @error = Canceled
+    end
+
+    def execute_or_skip
+      return if @executed
+
+      @pool.with_connection do |connection|
+        return unless @mutex.try_lock
+        begin
+          return if @executed
+          execute_query(connection, background: true)
+        ensure
+          @mutex.unlock
+        end
+      end
+    end
+
+    def result!
+      execute_or_wait
+      if @error
+        raise @error
+      else
+        @result
+      end
+    end
+
+    private
+      def execute_or_wait
+        return if @executed
+        if @mutex.try_lock
```

Comment:
This case feels like it's missing a second check of `@executed`. Maybe we could replace the whole method with a single `synchronize`?

Reply:
> This case feels like it's missing a second check of `@executed`

Good point, there indeed a very small race-cond window here.

> Maybe we could replace the whole method with a single synchronize?

I don't think so. `synchronize` would block until the mutex can be aquired. I use `try_lock` on purpose here to know wether the query is being executed by a background thread, or wether it is still queued.

But yes, there might be ways to refactor this a bit.

Reply:
I think that's needed in `execute_or_skip` above, but here in the else case (when the lock is held) this is calling `@mutex.synchronize` anyways, so I _think_ we can use just the one.